# Chapter 6: Loops and Repetitive Operations

**B.Pharm companion Jupyter notebook — worked examples and concept demonstrations**

## Chapter focus

Use for and while loops, range(), counters, accumulators, nested loops, break, continue and sentinel-controlled input.

## Pharmacy context

Repeated laboratory-style observations are summarized using counts and totals. The acceptance bands and stopping rules are simulated; they are not pharmacopoeial specifications.

## Important notes and how to use this notebook

1. Use a Python 3 kernel. Read the problem and algorithm, predict the result, then run the cell with **Shift+Enter**.
2. Choose **Restart Kernel and Run All Cells** to reproduce the saved outputs. Run from top to bottom; later demonstrations may use earlier variables.
3. Outputs below code cells were produced by execution. Paths, versions, memory diagnostics and set display order can differ between computers.
4. All records are simulated for education. Retain units and distinguish a descriptive result from a clinical or regulatory decision.
5. Files written by examples are kept in `outputs/chapter_06`. Rerunning a file-writing example replaces its own generated output.
6. Input examples replay the printed sample entries by default. `classroom_input()` is a small teaching helper around Python's `input()`; select interactive mode to enter your own values.

These examples use the Python standard library; no external dataset is needed.


In [1]:
from pathlib import Path

WORKSHOP_ROOT = Path.cwd().resolve()
OUTPUT_DIR = WORKSHOP_ROOT / "outputs" / "chapter_06"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

import builtins
INTERACTIVE_INPUT = False

def classroom_input(prompt=""):
    """Read real input or replay this example's documented sample entries."""
    if INTERACTIVE_INPUT:
        return builtins.input(prompt)
    try:
        answer = next(_sample_answers)
    except StopIteration as error:
        raise RuntimeError("Sample entries exhausted. Review the algorithm or use interactive mode.") from error
    print(prompt + answer)
    return answer

print("Chapter 6 ready. Generated files: outputs/chapter_06")

Chapter 6 ready. Generated files: outputs/chapter_06


## Contents

- [01. Worked Example 6.1 - Display Sample Numbers Using a for Loop](#example-6-1)
- [02. Worked Example 6.2 - Generate Dissolution Time Points](#example-6-2)
- [03. Worked Example 6.3 - Countdown of Samples Remaining](#example-6-3)
- [04. Worked Example 6.4 - Count Simulated Assay Results Meeting a Classroom Criterion](#example-6-4)
- [05. Worked Example 6.5 - Accumulate Tablet Weights and Compute a Mean](#example-6-5)
- [06. Worked Example 6.6 - Batch and Replicate Labels](#example-6-6)
- [07. Worked Example 6.7 - Stop When a Temperature Excursion Is Detected](#example-6-7)
- [08. Worked Example 6.8 - Skip a Missing Measurement](#example-6-8)
- [09. Worked Example 6.9 - Count ADR Classroom Entries Until END](#example-6-9)
- [10. 6.17 Integrated Worked Program - Small QC Summary](#example-6-10)

<a id="example-6-1"></a>

## 01. Worked Example 6.1 - Display Sample Numbers Using a for Loop

### Problem statement

Display laboratory sample numbers from 1 to 5.

### Objective

Use a for loop to repeat a simple output statement a known number of times.

### Pharmacy context

This pattern can label a small set of simulated laboratory samples, dissolution vessels or replicate observations during classroom exercises.

### Algorithm

1. Start.
2. Generate integers 1 through 5 using range(1, 6).
3. For each value, store it in sample_no.
4. Display the current sample number.
5. Stop after the range is exhausted.

### Code and executed output

The output appears directly below the code cell.

In [2]:
for sample_no in range(1, 6):
    print("Sample number:", sample_no)

Sample number: 1
Sample number: 2
Sample number: 3
Sample number: 4
Sample number: 5


### Explanation

range(1, 6) generates 1, 2, 3, 4 and 5. The stop value 6 is excluded. The indented print() statement runs once for each generated number.

### Interpretation

This pattern can label a small set of simulated laboratory samples, dissolution vessels or replicate observations during classroom exercises.

<a id="example-6-2"></a>

## 02. Worked Example 6.2 - Generate Dissolution Time Points

### Problem statement

Display simulated dissolution sampling times at 0, 5, 10, 15, 20, 25 and 30 minutes.

### Objective

Use the step argument of range() to generate equally spaced time points.

### Pharmacy context

Regular time intervals are common in simulated concentration-time and dissolution exercises. The program demonstrates how scheduled observation points can be generated without typing every value manually.

### Algorithm

1. Start at 0 minutes.
2. Use a step of 5 minutes.
3. Continue until 30 minutes is included.
4. Display each time point.

### Code and executed output

The output appears directly below the code cell.

In [3]:
for time_min in range(0, 31, 5):
    print(time_min, "minutes")

0 minutes
5 minutes
10 minutes
15 minutes
20 minutes
25 minutes
30 minutes


### Explanation

The stop argument is 31 because range() excludes the stop value. A step of 5 creates regular five-minute intervals.

### Interpretation

Regular time intervals are common in simulated concentration-time and dissolution exercises. The program demonstrates how scheduled observation points can be generated without typing every value manually.

<a id="example-6-3"></a>

## 03. Worked Example 6.3 - Countdown of Samples Remaining

### Problem statement

A classroom activity has five samples left to inspect. Display the number remaining until all five have been processed.

### Objective

Demonstrate initialization, condition testing and update in a while loop.

### Pharmacy context

A countdown loop can represent a controlled repeated workflow, such as completing a fixed queue of simulated samples.

### Algorithm

1. Set samples_remaining to 5.
2. While samples_remaining is greater than 0, display its value.
3. Decrease samples_remaining by 1.
4. When the condition becomes False, display a completion message.

### Code and executed output

The output appears directly below the code cell.

In [4]:
samples_remaining = 5

while samples_remaining > 0:
    print("Samples remaining:", samples_remaining)
    samples_remaining -= 1

print("Inspection sequence complete")

Samples remaining: 5
Samples remaining: 4
Samples remaining: 3
Samples remaining: 2
Samples remaining: 1
Inspection sequence complete


### Explanation

The variable decreases by 1 after every iteration. When it reaches 0, the condition samples_remaining > 0 becomes False and the loop ends.

### Interpretation

A countdown loop can represent a controlled repeated workflow, such as completing a fixed queue of simulated samples.

<a id="example-6-4"></a>

## 04. Worked Example 6.4 - Count Simulated Assay Results Meeting a Classroom Criterion

### Problem statement

Enter five simulated assay percentages and count how many fall within the classroom band 95.0% to 105.0%.

### Objective

Use a counter with a for loop and a conditional statement.

### Pharmacy context

The example illustrates how repeated QC-like observations can be classified and counted. The numerical band is explicitly a classroom value for programming practice.

### Algorithm

1. Set accepted_count to 0.
2. Repeat five times.
3. Read assay_percent.
4. If 95.0 <= assay_percent <= 105.0, increase accepted_count by 1.
5. After the loop, display accepted_count.

### Important notes

Saved output uses these sample entries in order: `99.2`, `94.6`, `101.1`, `105.4`, `98.7`. Set INTERACTIVE_INPUT = True in the setup cell to type your own entries.

### Code and executed output

The output appears directly below the code cell.

In [5]:
_sample_answers = iter(['99.2', '94.6', '101.1', '105.4', '98.7'])

accepted_count = 0

for sample_no in range(1, 6):
    assay_percent = float(classroom_input(f"Enter assay % for sample {sample_no}: "))
    if 95.0 <= assay_percent <= 105.0:
        accepted_count += 1

print("Results inside classroom band:", accepted_count)

Enter assay % for sample 1: 99.2
Enter assay % for sample 2: 94.6
Enter assay % for sample 3: 101.1
Enter assay % for sample 4: 105.4
Enter assay % for sample 5: 98.7
Results inside classroom band: 3


### Explanation

accepted_count changes only when the condition is True. It is therefore a counter, not a running total of the assay values.

### Interpretation

The example illustrates how repeated QC-like observations can be classified and counted. The numerical band is explicitly a classroom value for programming practice.

<a id="example-6-5"></a>

## 05. Worked Example 6.5 - Accumulate Tablet Weights and Compute a Mean

### Problem statement

Enter the weights of four simulated tablets and calculate the total and mean weight.

### Objective

Use an accumulator to add repeated numerical values.

### Pharmacy context

Accumulation is a basic pattern used before calculating totals and averages of repeated measurements. The data shown are simulated.

### Algorithm

1. Set total_weight_mg to 0.
2. Repeat four times.
3. Read weight_mg.
4. Add weight_mg to total_weight_mg.
5. Divide total by 4 to obtain the mean.
6. Display total and mean.

### Important notes

Saved output uses these sample entries in order: `501`, `498`, `503`, `500`. Set INTERACTIVE_INPUT = True in the setup cell to type your own entries.

### Code and executed output

The output appears directly below the code cell.

In [6]:
_sample_answers = iter(['501', '498', '503', '500'])

total_weight_mg = 0.0
number_of_tablets = 4

for tablet_no in range(1, number_of_tablets + 1):
    weight_mg = float(classroom_input(f"Enter weight of tablet {tablet_no} (mg): "))
    total_weight_mg += weight_mg

mean_weight_mg = total_weight_mg / number_of_tablets
print("Total weight (mg):", total_weight_mg)
print("Mean weight (mg):", mean_weight_mg)

Enter weight of tablet 1 (mg): 501
Enter weight of tablet 2 (mg): 498
Enter weight of tablet 3 (mg): 503
Enter weight of tablet 4 (mg): 500
Total weight (mg): 2002.0
Mean weight (mg): 500.5


### Explanation

total_weight_mg is an accumulator because every entered weight is added to the running total. The mean is calculated only after all four values have been processed.

### Interpretation

Accumulation is a basic pattern used before calculating totals and averages of repeated measurements. The data shown are simulated.

<a id="example-6-6"></a>

## 06. Worked Example 6.6 - Batch and Replicate Labels

### Problem statement

Generate labels for 3 simulated batches, each having 2 replicate measurements.

### Objective

Demonstrate two-level repetition using nested for loops.

### Pharmacy context

Nested loops are useful when classroom data have two repeated dimensions, such as batches and replicates, formulations and time points, or rows and columns of a small plate layout.

### Algorithm

1. Repeat batch_no from 1 to 3.
2. For each batch, repeat replicate_no from 1 to 2.
3. Display the batch-replicate combination.
4. Stop after all combinations are produced.

### Code and executed output

The output appears directly below the code cell.

In [7]:
for batch_no in range(1, 4):
    for replicate_no in range(1, 3):
        print("Batch", batch_no, "- Replicate", replicate_no)

Batch 1 - Replicate 1
Batch 1 - Replicate 2
Batch 2 - Replicate 1
Batch 2 - Replicate 2
Batch 3 - Replicate 1
Batch 3 - Replicate 2


### Explanation

The inner loop executes twice for each of the three outer-loop iterations, so the print() statement runs 3 × 2 = 6 times.

### Interpretation

Nested loops are useful when classroom data have two repeated dimensions, such as batches and replicates, formulations and time points, or rows and columns of a small plate layout.

<a id="example-6-7"></a>

## 07. Worked Example 6.7 - Stop When a Temperature Excursion Is Detected

### Problem statement

Up to six simulated storage-temperature readings will be entered. Stop processing immediately if a value exceeds 8.0 °C.

### Objective

Use break to terminate a loop when a stop condition is met.

### Pharmacy context

The pattern demonstrates early termination when a classroom-defined alert condition is detected. Actual storage specifications must come from approved product and regulatory requirements.

### Algorithm

1. Repeat for a maximum of six readings.
2. Enter temperature_c.
3. If temperature_c > 8.0, display an excursion message and break.
4. Otherwise display that the reading was processed.
5. Continue until six readings are processed or break occurs.

### Important notes

This specific example checks only an upper threshold. It does not implement a complete lower-and-upper storage-range screen.

Saved output uses these sample entries in order: `5.1`, `6.0`, `8.6`. Set INTERACTIVE_INPUT = True in the setup cell to type your own entries.

### Code and executed output

The output appears directly below the code cell.

In [8]:
_sample_answers = iter(['5.1', '6.0', '8.6'])

for reading_no in range(1, 7):
    temperature_c = float(classroom_input(f"Reading {reading_no} (°C): "))

    if temperature_c > 8.0:
        print("Classroom excursion flag - stopping input")
        break

    print("Reading processed")

Reading 1 (°C): 5.1
Reading processed
Reading 2 (°C): 6.0
Reading processed
Reading 3 (°C): 8.6
Classroom excursion flag - stopping input


### Explanation

When the third value meets the stop condition, break exits the for loop. Readings 4 to 6 are not requested.

### Interpretation

The pattern demonstrates early termination when a classroom-defined alert condition is detected. Actual storage specifications must come from approved product and regulatory requirements.

<a id="example-6-8"></a>

## 08. Worked Example 6.8 - Skip a Missing Measurement

### Problem statement

Enter five simulated pH readings. Use -1 as a classroom code for a missing reading. Skip missing values and add only available readings to the total.

### Objective

Use continue to ignore a selected iteration without terminating the loop.

### Pharmacy context

Missing-data handling should be explicit. This classroom example shows how a program can skip a designated missing marker while preserving the repeated workflow.

### Algorithm

1. Set total_ph and valid_count to 0.
2. Repeat five times.
3. Read ph_value.
4. If ph_value == -1, display a missing message and continue.
5. Otherwise add ph_value to total_ph and increase valid_count.
6. After the loop, display valid_count and total_ph.

### Important notes

The sum of the pH numbers demonstrates an accumulator only; pH is logarithmic, so this sum is not a scientifically meaningful combined pH measurement.

Saved output uses these sample entries in order: `6.8`, `-1`, `7.0`, `6.9`, `7.1`. Set INTERACTIVE_INPUT = True in the setup cell to type your own entries.

### Code and executed output

The output appears directly below the code cell.

In [9]:
_sample_answers = iter(['6.8', '-1', '7.0', '6.9', '7.1'])

total_ph = 0.0
valid_count = 0

for reading_no in range(1, 6):
    ph_value = float(classroom_input(f"pH reading {reading_no} (-1 if missing): "))

    if ph_value == -1:
        print("Missing reading skipped")
        continue

    total_ph += ph_value
    valid_count += 1

print("Valid readings:", valid_count)
print("Total of valid pH readings:", total_ph)

pH reading 1 (-1 if missing): 6.8
pH reading 2 (-1 if missing): -1
Missing reading skipped
pH reading 3 (-1 if missing): 7.0
pH reading 4 (-1 if missing): 6.9
pH reading 5 (-1 if missing): 7.1
Valid readings: 4
Total of valid pH readings: 27.800000000000004


### Explanation

The continue statement prevents the missing code -1 from being included in the accumulator or count. The loop itself continues with the next reading.

### Interpretation

Missing-data handling should be explicit. This classroom example shows how a program can skip a designated missing marker while preserving the repeated workflow.

<a id="example-6-9"></a>

## 09. Worked Example 6.9 - Count ADR Classroom Entries Until END

### Problem statement

Accept simulated ADR category entries until the user types END. Count how many non-empty entries are recorded.

### Objective

Use a while loop controlled by a text sentinel.

### Pharmacy context

The example models open-ended data-entry logic using simulated ADR labels. It is a programming exercise, not a pharmacovigilance reporting system.

### Algorithm

1. Set report_count to 0.
2. Read adr_entry.
3. While adr_entry is not END, process the entry.
4. If the entry is not blank, increase report_count.
5. Read the next entry.
6. When END is entered, leave the loop and display report_count.

### Important notes

Saved output uses these sample entries in order: `nausea`, `rash`, `dizziness`, `END`. Set INTERACTIVE_INPUT = True in the setup cell to type your own entries.

### Code and executed output

The output appears directly below the code cell.

In [10]:
_sample_answers = iter(['nausea', 'rash', 'dizziness', 'END'])

report_count = 0
adr_entry = classroom_input("Enter ADR category or END: ").strip()

while adr_entry.upper() != "END":
    if adr_entry != "":
        report_count += 1
    adr_entry = classroom_input("Enter ADR category or END: ").strip()

print("Classroom entries recorded:", report_count)

Enter ADR category or END: nausea
Enter ADR category or END: rash
Enter ADR category or END: dizziness
Enter ADR category or END: END
Classroom entries recorded: 3


### Explanation

The loop does not need to know the number of entries beforehand. It stops only when END is entered. strip() removes surrounding spaces and upper() makes END detection case-insensitive.

### Interpretation

The example models open-ended data-entry logic using simulated ADR labels. It is a programming exercise, not a pharmacovigilance reporting system.

<a id="example-6-10"></a>

## 10. 6.17 Integrated Worked Program - Small QC Summary

### Problem statement

Enter a user-selected number of simulated assay results. Skip impossible classroom entries outside 0 to 200%, count valid results, count results inside the classroom band 95 to 105%, and calculate the mean of valid results.

### Objective

Integrate loop control, validation, counting and accumulation in one beginner-level program.

### Pharmacy context

This program demonstrates the logic of processing repeated laboratory-style observations, separating invalid data from analyzable data, summarizing the valid values and counting values in a simulated classroom band.

### Algorithm

1. Read number_of_samples.
2. Initialize total_assay, valid_count and accepted_count to 0.
3. Repeat for each planned sample.
4. Read assay_percent.
5. If assay_percent is outside 0 to 200, display a message and continue.
6. Add valid assay_percent to total_assay and increase valid_count.
7. If 95 <= assay_percent <= 105, increase accepted_count.
8. After the loop, calculate mean_assay only when valid_count > 0.
9. Display the summary.

### Important notes

Saved output uses these sample entries in order: `5`, `99.2`, `102.0`, `-3`, `94.8`, `101.5`. Set INTERACTIVE_INPUT = True in the setup cell to type your own entries.

### Code and executed output

The output appears directly below the code cell.

In [11]:
_sample_answers = iter(['5', '99.2', '102.0', '-3', '94.8', '101.5'])

number_of_samples = int(classroom_input("How many samples? "))

total_assay = 0.0
valid_count = 0
accepted_count = 0

for sample_no in range(1, number_of_samples + 1):
    assay_percent = float(classroom_input(f"Assay % for sample {sample_no}: "))

    if assay_percent < 0 or assay_percent > 200:
        print("Invalid classroom value skipped")
        continue

    total_assay += assay_percent
    valid_count += 1

    if 95.0 <= assay_percent <= 105.0:
        accepted_count += 1

if valid_count > 0:
    mean_assay = total_assay / valid_count
    print("Valid results:", valid_count)
    print("Inside classroom band:", accepted_count)
    print("Mean assay %:", round(mean_assay, 2))
else:
    print("No valid results were entered")

How many samples? 5
Assay % for sample 1: 99.2
Assay % for sample 2: 102.0
Assay % for sample 3: -3
Invalid classroom value skipped
Assay % for sample 4: 94.8
Assay % for sample 5: 101.5
Valid results: 4
Inside classroom band: 3
Mean assay %: 99.38


### Explanation

The for loop controls the planned repetitions. continue skips an invalid classroom entry before it reaches the accumulator. total_assay adds valid values, valid_count records how many valid values were processed, and accepted_count records how many meet the simulated band. The final if statement prevents division by zero.

### Interpretation

This program demonstrates the logic of processing repeated laboratory-style observations, separating invalid data from analyzable data, summarizing the valid values and counting values in a simulated classroom band.

## Review and practice

- Change one simulated input and predict the effect before rerunning the relevant cell.
- State the unit and meaning of each reported result.
- Explain one input condition or limitation that matters for the example.
- Restart the kernel and run all cells to check that the notebook is reproducible.

## Source and validation

Adapted from **Chapter_6_Loops_and_Repetitive_Operations_BPharm(2).docx**. Worked programs and the complete code demonstrations are retained in chapter order. Installation commands are optional instructions; intentionally broken debugging examples are shown as corrected solutions. Notebook-specific changes are recorded in `Notebook_Guide_and_Validation.md`.